### PR TITLE
URI repository paths

### LINKED ISSUE #1908: Version Repository URI Paths
**Goal**

To support URI schema paths for the versions repository.

**Motivation**

Currently only local paths are support, but there is not reason why Openpype couldn't download versions from cloud storages.

Our use case for this is that we have a project where we cant guarantee that the versions repository path is the same for remote users, but we can provide a download link to a dropbox folder which would be the same for everyone.

**Implementation**

We would use URI paths to define which type of storage the version repository is on:

`file://` > local file system
`git://` > git repository
`dropbox://` > dropbox folder/link

For backwards compatibility any paths that fail or do not return a URI scheme, should be treated as local file system path.

### PR DESCRIPTION
**Pull request by tokejepsen**
**Goal**

Support URI paths like `file://` and `dropbox://`.

Closes #1908 

**Motivation**

For some of our remote users we cant ensure access to local repository paths, so we have a need for storing our versions in the cloud.

**Steps to test**
Dropbox is the only supported cloud storage atm. For Dropbox you'll need to create an app: https://developers.dropbox.com/oauth-guide

I opted to use an app because you can have a separate root folder for the app, which makes it easier to organize.

Input the access token in the settings `Dropbox` section.

**Implementation**

I initially tried to do custom implementations of Path like objects, but quickly found that the library https://cloudpathlib.drivendata.org/ had already figured things out, so I opted to use that.
The library does not however support Dropbox (which we need), so there is a custom class for Dropbox. That same goes for the uri `file://` which is just a wrapper for the built-in Path object.
The `AnyPath` wrapper determines which class to use depending on the uri and if its a local path it'll revert to built-in Path object. This means we can use the `AnyPath` as a replacement for `Path` everywhere.

I've initially put its all in `tools` but it might need a separate module.

There is a caching system in `cloudpath` that'll localize the data when reading it. In order for this to work smoothly with the user data directory, I'm using the user data directory as the cache directory. To optimize download I changed to keeping the zip file after extraction. To further optimize download I've changed the validation to happen after the versions have been found, else we would be downloading zip files just validate against the `version.py` inside.
To keep the user data directory clean, we delete any zip file that is not the current and any directory. Let me know if there is any issues with this, that you can think of.

The only thing I dont like about this implementation is the hack to get the system settings inside of the Dropbox class. This is because when we get the repository paths and validate their existence, we need the dropbox token.

[cuID:yjx9fd]

### CODE REVIEW

**Comment by mkolar** [comment]
resolves PYPE-1290

**Comment by tokejepsen** [comment]
We could potentially unify the setup process with https://github.com/pypeclub/OpenPype/pull/1979, by using Team Folders, so people only have to make one Dropbox app, and one place for all the settings.

**Review by antirotor — CHANGES_REQUESTED** [review]
I've just found one unpleasant side-effect of this - if you enable dropbox module, set your token and then enter path to OpenPype repositories to `dropbox://...` and it will fail for some reason, you can't run settings anymore (anything from openpype in fact) as `get_openpype_path_from_db()` is called early in bootstrap process and it will pull in dropbox to check if path exists.

Some solution is with changing order how path is determined here:
https://github.com/pypeclub/OpenPype/blob/2a01fb8f3165b237161f2ad6877d9a6213c79087/start.py#L811

so check for `OPENPYPE_PATH` should be before this, but nevertheless we should probably handle dropbox errors more gracefully.

**Comment by tokejepsen** [comment]
> but nevertheless we should probably handle dropbox errors more gracefully.

Do you have any stack trace of the errors?

**Comment by antirotor** [comment]
> Do you have any stack trace of the errors?

yeah, sorry:

```
Traceback (most recent call last):
  File "C:\Users\annat\Documents\pype\3.0\start.py", line 940, in <module>
    boot()
  File "C:\Users\annat\Documents\pype\3.0\start.py", line 786, in boot
    openpype_path = get_openpype_path_from_db(openpype_mongo)
  File "C:\Users\annat\Documents\pype\3.0\igniter\tools.py", line 590, in get_openpype_path_from_db
    if AnyPath(path).exists():
  File "C:\Users\annat\Documents\pype\3.0\.venv\lib\site-packages\cloudpathlib\cloudpath.py", line 302, in exists
    return self.client._exists(self)
  File "C:\Users\annat\Documents\pype\3.0\igniter\tools.py", line 278, in _exists
    self.client.files_get_metadata(cloud_path.path)
  File "C:\Users\annat\Documents\pype\3.0\.venv\lib\site-packages\dropbox\base.py", line 1512, in files_get_metadata
    None,
  File "C:\Users\annat\Documents\pype\3.0\.venv\lib\site-packages\dropbox\dropbox_client.py", line 329, in request
    timeout=timeout)
  File "C:\Users\annat\Documents\pype\3.0\.venv\lib\site-packages\dropbox\dropbox_client.py", line 485, in request_json_string_with_retry
    timeout=timeout)
  File "C:\Users\annat\Documents\pype\3.0\.venv\lib\site-packages\dropbox\dropbox_client.py", line 611, in request_json_string
    raise AuthError(request_id, err)
dropbox.exceptions.AuthError: AuthError('f541561c6e5e4c7819a01325dfa470ce4', AuthError('missing_scope', TokenScopeError(required_scope='files.metadata.read')))
```

**Comment by tokejepsen** [comment]
Good point! We'll need to gracefully authenticate Dropbox else use the local path I guess?

**Comment by antirotor** [comment]
> Good point! We'll need to gracefully authenticate Dropbox else use the local path I guess?

Maybe it would be enough to catch exception in:
```python
if AnyPath(path).exists():
   ...
```
because you can  say that if  `exists()` fails, path doesn't exist :D

**Comment by tokejepsen** [comment]
True, we could do a try catch all.

And fallback on the local user directory?

**Comment by antirotor** [comment]
> And fallback on the local user directory?

Yeah, issue warning and fallback to what can be found locally. That warning should be in UI if not in headless mode so artist would know exactly that he's not using "official" studio version but possibly outdated one and he need to contact someone to fix it.

**Comment by tokejepsen** [comment]
Note to self. Something that keeps tripping me up when working with the Dropbox API is that the permissions are attached to the token. So if you update the permissions you'll need to update the token as well.

**Comment by tokejepsen** [comment]
This is ready for another review.

File: igniter/tools.py [hunk 1]
```diff
@@ -14,6 +15,305 @@
     ConfigurationError,
     OperationFailure
 )
+from cloudpathlib import AnyPath
```
Comment (iLLiCiTiT) [inline igniter/tools.py:18]: Code specific for dropbox should be in separated file and only code that is used here should be imported. There is a lot of code which is dropbox specific but does not make sense in context of `igniter/tools` for example class `FilePath` is dropbox specific but I this context it is confusing.

**Comment by ClementHector** [comment]
I will soon start implementing the same for Google Drive. Wouldn't the easiest way be to fork cloudpathlib and add the class for `
`dropbox` and `gdrive` to it?

**Comment by tokejepsen** [comment]
> I will soon start implementing the same for Google Drive. Wouldn't the easiest way be to fork cloudpathlib and add the class for
dropbox and gdrive to it?

Yeah, it might be. I just didnt have time to get figure out where it should be. Also a fork might be more trouble managing with OP than a custom class, but forking might lead to a PR to origin repository which would be better.

**Comment by mkolar** [comment]
We had to actually put this on hold, until we finish refactor of the repository. To implement this cleanly it should be modules that provide the path cloud location, however those are not loaded until we start OP. there's a bit of an chicken and egg situation. We're counting on separating these responsibilities properly soon.

**Comment by mkolar** [comment]
Unfortunatelly, we won't be able to merge this in this state. It is however one of the first things to tackle as soon as we're done with the refactor of the client application, which is being changed to allow for this to be re-implemented a lot more robustly, without having to hardcode any specific providers into the product.

**Pull request closed without merging**